aten/src/ATen/native/native_functions.yaml

```diff
@@ -1399,15 +1399,23 @@
   variants: function
 
 
+# NB: The function overloads are removed to avoid a nasty bug where
+# you say at::native_sparse_coo_tensor(indices, values), and then
+# it does the dispatch based on indices (wrong wrong wrong!)  Without
+# the variants, you must call this on type directly.
+#
+# Maybe this meant we were supposed to take a dtype as an argument here.
+# Hmmmmm.
```

Comment:
That's not unreasonable, although it wouldn't be enough for this case, because neither argument is the actual type you want the dispatch to happen on in the end.

Actually, when C10 dispatcher lands none of this will matter because you won't have to pick an argument to dispatch on.

Reply:
Also I think just taking a Type or ScalarType would work (as you suggest) -- you can do all the same logic that is implemented in tensor_new here.

Reply:
I attempted this patch, but I triggered a codegen bug which I don't feel like fixing:

```
In file included from torch/csrc/autograd/generated/python_torch_functions.cpp:23:0:
torch/csrc/autograd/generated/python_torch_functions_dispatch.h: In function ‘at::Tensor torch::autograd::dispatch__native_sparse_co
o_tensor_unsafe(const at::Tensor&, const at::Tensor&, at::IntList)’:
torch/csrc/autograd/generated/python_torch_functions_dispatch.h:584:25: error: ‘dtype’ was not declared in this scope
   maybe_initialize_cuda(dtype);
```

Reply:
https://github.com/ezyang/pytorch-1/tree/snapshot/taking-type-for-sparse-coo-tensor